src/Lib.P2P/SwarmService.cs

```diff
@@ -0,0 +1,1063 @@
+               .SelectMany(a => a)
+               .Where(a => !blackList.Contains(a.WithoutPeerId()))
+               .Select(a => a.WithPeerId(remote.Id))
+               .Distinct()
+               .ToArray();
+            if (possibleAddresses.Length == 0) throw new Exception($"{remote} has no known or reachable address.");
+
+            // Try the various addresses in parallel.  The first one to complete wins.
+            PeerConnection connection = null;
+            try
+            {
+                using (var timeout = new CancellationTokenSource(TransportConnectionTimeout))
+                using (var cts = CancellationTokenSource.CreateLinkedTokenSource(timeout.Token, cancel))
+                {
+                    var attempts = possibleAddresses
+                       .Select(a => DialAsync(remote, a, cts.Token));
+                    connection = await TaskHelper.WhenAnyResultAsync(attempts, cts.Token).ConfigureAwait(false);
+                    cts.Cancel(); // stop other dialing tasks.
+                }
+            }
+            catch (Exception e)
+            {
+                var attemped = string.Join(", ", possibleAddresses.Select(a => a.ToString()));
+                log.Trace($"Cannot dial {attemped}");
+                throw new Exception($"Cannot dial {remote}.", e);
+            }
+
+            // Do the connection handshake.
+            try
+            {
+                MountProtocols(connection);
+                IEncryptionProtocol[] security = null;
+                lock (protocols)
+                {
+                    security = protocols.OfType<IEncryptionProtocol>().ToArray();
+                }
+
+                await connection.InitiateAsync(security, cancel).ConfigureAwait(false);
+                await connection.MuxerEstablished.Task.ConfigureAwait(false);
+                Identify1 identify = null;
+                lock (protocols)
+                {
+                    identify = protocols.OfType<Identify1>().First();
+                }
+
+                await identify.GetRemotePeerAsync(connection, cancel).ConfigureAwait(false);
+            }
+            catch (Exception)
+            {
+                connection.Dispose();
+                throw;
+            }
+
+            var actual = Manager.Add(connection);
+            if (actual == connection) ConnectionEstablished?.Invoke(this, connection);
+
+            return actual;
+        }
+
+        private async Task<PeerConnection> DialAsync(Peer remote, MultiAddress addr, CancellationToken cancel)
+        {
+            // TODO: HACK: Currenty only the ipfs/p2p is supported.
+            // short circuit to make life faster.
+            if (addr.Protocols.Count != 3
+             || !(addr.Protocols[2].Name == "ipfs" || addr.Protocols[2].Name == "p2p"))
+                throw new Exception($"Cannnot dial; unknown protocol in '{addr}'.");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)